### ISSUE TITLE
Method signatures containing references to properties of constrained generic types are resolving through the constraint instead of the supplied type argument

### ISSUE THREAD

**Issue by rotemdan** [Microsoft/TypeScript#12651]
**TypeScript Version:** nightly (2.2.0-dev.20161203)

I currently find it somewhat hard to precisely describe the issue/suggestion, but I believe this use case should demonstrate it clearly:

```ts
type MethodDescriptor = {
	name: string;
	args: any[];
	returnValue: any;
}

function dispatchMethod<M extends MethodDescriptor>(name: M['name'], args: M['args']): M['returnValue'];

type SomeMethodDescriptor = {
	name: "someMethod";
	args: [number, string];
	returnValue: string[];
}

// Currently, there's no compilation error on the the invalid second argument type and the 
// return type is inferred as "any" instead of "string[]":
let result = dispatchMethod<SomeMethodDescriptor>("someMethod", ["hello", 35]);
```

The general idea here is that the types are used unconventionally: not to describe interfaces to run-time objects but as a way to "pack" a set of type arguments into an "object-like" compile-time entity. Perhaps this could be called something like "packed type arguments" or "type argument classes" or maybe "type argument schema containers"? I'm not sure.

Anyway, I was trying to write a general purpose dispatcher that can invoke a set of actions, and that seemed like a more elegant and type safe alternative to the more conventional:
```ts
function dispatchMethod<A extends any[], R>(name: string, args: A): R;

dispatchMethod<[number, string], string[]>("someMethod", ["hello", 35]); // <- errors as expected
```

**Comment by rotemdan** [comment]
A convenient way to illustrate this is that:

```ts
type MethodDescriptor = {
	name: string;
	args: any[];
	returnValue: any;
}
```
Represent a set of base constraints to a **compile-time** "object-like" entity containing type arguments, e.g.

1. Type of `name` must be assignable to `string`.
2. Type of `args` must be assignable to `any[]`.
3. Type of `returnValue` must be assignable to `any`.

And:
```ts
type SomeMethod = {
	name: "someMethod";
	args: [number, string];
	returnValue: string[];
}
```
Is an "instance" of that "class", in the sense that it associates more specific **types** to the properties.

Now:
```ts
function dispatchMethod<M extends MethodDescriptor>(name: M['name'], args: M['args']): M['returnValue'];
```
References the **types** of properties on the type argument `M`.

Note that in this example both the types `MethodDescriptor` and `SomeMethod` are **never actually applied to any concrete run-time object**.

I know this may be seen as somewhat abstract so I hope this helps..

**Comment by rotemdan** [comment]
In combination with discriminated unions. It is possible to encapsulate the dispatcher in class such that only one "type schema" is actually needed for all supported actions:

```ts
type MethodDescriptor = {
	name: string;
	args: any[];
	returnValue: any;
}

class Dispatcher<M extends MethodDescriptor> {
	dispatch(name: M['name'], args: M['args']): M['returnValue'] {
		// ...
	}
}

type StoreMethodDescriptor = {
	name: "store";
	args: [string, number[], boolean];
	returnValue: void;
}

type OpenMethodDescriptor = {
	name: "open";
	args: [string];
	returnValue: void;
}

let dispatcher = new Dispatcher<StoreMethodDescriptor | OpenMethodDescriptor>();

// The correct argument types are inferred and validated automatically using the string literal 
// as discriminant:
dispatcher.dispatch("store", ["mydata.txt", [1, 2, 3, 4], true]);
```

**Comment by rotemdan** [comment]
I realized there's a subtle assumption I introduced here in my union example.

Consider this case:
```ts
type MyType = { a: string; b: number } | { a: boolean; b: string[] };

function test(x: MyType['a'], y: MyType['b']) {}
test("hi", ["a", "b", "c"]); // <- currently doesn't error
```
Here, `MyType['a']` is currently inferred as `string | boolean` and `MyType['b']` is currently inferred as `number | string[]`, so `test("hi", ["a", "b", "c"])` currently doesn't error.

Is this the right way to go? maybe? However, the case I described was slightly different:
```ts
class Test<T extends { a: any; b: any }> {
	test(x: T['a'], y: T['b']) {
		// ...
	}
}
const instance = new Test<{ a: string; b: number } | { a: boolean; b: string[] }>;
```
Should both `T['a']` and `T['b']` be required to be consistent with a unique member of the union? I'm not completely sure. This could only be seen as a slight ambiguity of the syntax though. Perhaps if there was a helper syntax to "constrain" a type to a single member of a union, something like `uniquely extends` :

```ts
class Test<T extends { a: any; b: any }> {
	test<M uniquely extends T>(x: M['a'], y: M['b']) {
		// ...
	}
}
const instance = new Test<{ a: string; b: number } | { a: boolean; b: string[] }>;
```
The suggested keyword might not be great though.. perhaps there's a better one..

**Comment by rotemdan** [comment]
A simpler alternative to `uniquely extends` is to use something like `specifies` instead of `extends` for the main type parameter:

```ts
class Test<T specifies { a: any; b: any }> {
	test(x: T['a'], y: T['b']) {
		// ...
	}
}

const instance = new Test<{ a: string; b: number } | { a: boolean; b: string[] }>();

instance.test("hi", ["a", "b", "c"]); // <-- Errors due to the semantics of "specifies".
```
Although `specifies` can be a bit ambiguous as it might suggest the type argument _can't_ be a union (in the sense it must be "specific"), I think the potential confusion is only minor, and the reduced complexity and increased user friendliness may be worth it.

Another advantage of `specifies` is that it may support future syntax that wouldn't be appropriate for `extends`. Anyway, the name best name I found so far for this is concept is _type schemas_, so `specifies` seems to go well with that.

**Comment by ahejlsberg** [comment]
The issue described in the original post is now fixed by #12770.

**Comment by rotemdan** [comment]
@ahejlsberg 

Thanks! I'll grab the nightly build tomorrow and see if I can start making use of this new inference capability in my code. 

If you find any of these additional ideas I mentioned here interesting feel free to make use of them in the future. They were originally meant as motivating examples. I'm not sure, though, whether I'm in a good position to figure out a whole proposal from them at this point. It would be interesting, perhaps, to first see what other people would do / want to do with these inference patterns and then try to come to some form of understanding what's the 'best' way to approach this.
